File: canonical/swift/Sources/YXProtocol/Transport/RateLimiter.swift
import Foundation

/// Rate limiting using sliding window per peer
///
/// ⚠️ CRITICAL DEFAULT: maxRequests = 10,000 (SDTS Issue #2)
/// Swift v1.0.2 used 100 req/60s while Python used 10,000, which blocked
/// high-frequency trading. The default MUST be 10,000.
actor RateLimiter {

    private let maxRequests: Int
    private let windowSeconds: TimeInterval
    private var peerRequests: [String: [Date]] = [:]

    init(maxRequests: Int = 10_000, windowSeconds: TimeInterval = 60.0) {
        self.maxRequests = maxRequests
        self.windowSeconds = windowSeconds

        if maxRequests < 10_000 {
            print("⚠️  WARNING: RateLimiter maxRequests (\(maxRequests)) is below recommended 10,000")
            print("   This may block high-frequency trading systems (see SDTS Issue #2)")
        }
    }

    /// Check if request is allowed for peer (sliding window).
    func checkRateLimit(peerID: String, sourceAddr: String) -> Bool {
        let now = Date()
        let cutoff = now.addingTimeInterval(-windowSeconds)

        let key = peerID.isEmpty ? sourceAddr : peerID
        var requests = peerRequests[key, default: []]
        requests = requests.filter { $0 > cutoff }

        if requests.count >= maxRequests {
            peerRequests[key] = requests
            return false  // RATE LIMIT EXCEEDED
        }

        requests.append(now)
        peerRequests[key] = requests
        return true
    }

    /// Get current request count for peer in the window.
    func getCurrentCount(peerID: String) -> Int {
        let now = Date()
        let cutoff = now.addingTimeInterval(-windowSeconds)
        guard let requests = peerRequests[peerID] else {
            return 0
        }
        return requests.filter { $0 > cutoff }.count
    }

    func clear() {
        peerRequests.removeAll()
    }

    func getConfig() -> (maxRequests: Int, windowSeconds: TimeInterval) {
        return (maxRequests, windowSeconds)
    }
}

File: src/yx/transport/rate_limiter.py
"""
Rate limiting using sliding window per peer.

Traceability:
- protocol/specs/architecture/security-architecture.md (Layer 3: Rate Limiting)
- protocol/specs/technical/default-values.md (max_requests = 10,000 NOT 100!)
- SDTS Issue #2: Rate limiter default mismatch (was 100, must be 10,000)
"""

import time
from dataclasses import dataclass, field
from typing import Dict, List, Set
import logging

logger = logging.getLogger(__name__)


@dataclass
class RateLimiter:
    """
    Per-peer rate limiter with sliding window.

    CRITICAL: max_requests = 10,000 (for high-frequency trading support).
    See SDTS Issue #2 for consequences of wrong default.

    Traceability:
    - protocol/specs/architecture/security-architecture.md (Rate Limiting)
    - protocol/specs/technical/default-values.md (max_requests, rate_limit_window)
    """
    max_requests: int = 10000  # CRITICAL: 10,000 NOT 100
    window_seconds: float = 60.0
    trusted_guids: Set[str] = field(default_factory=set)
    _history: Dict[str, List[float]] = field(default_factory=dict)

    def check_rate_limit(self, peer_id: str, source_addr: tuple) -> bool:
        """
        Check if peer is within rate limit.

        Returns:
            True = allowed
            False = blocked (exceeded limit)
        """
        if peer_id in self.trusted_guids:
            return True

        now = time.time()
        if peer_id not in self._history:
            self._history[peer_id] = []

        history = self._history[peer_id]
        cutoff = now - self.window_seconds
        history[:] = [t for t in history if t > cutoff]

        if len(history) >= self.max_requests:
            logger.warning(f"Rate limit exceeded for {peer_id} from {source_addr}")
            return False

        history.append(now)
        return True

    def add_trusted_guid(self, guid_hex: str):
        self.trusted_guids.add(guid_hex.upper())

    def remove_trusted_guid(self, guid_hex: str):
        self.trusted_guids.discard(guid_hex.upper())

    def is_trusted(self, guid_hex: str) -> bool:
        return guid_hex.upper() in self.trusted_guids

    def reset_peer(self, peer_id: str):
        self._history.pop(peer_id, None)
